### PR TITLE
feat(compass-assistant): add GuideCue support to DrawerSection COMPASS-9795

### PR DESCRIPTION
**Pull request by lerouxb**
This is a temporary hack until we get better support for targeting the toolbar icon in leafygreen. The overall interface on <DrawerSection> should survive and the dom element lookups should hopefully go away.

To test locally you can run this in the console and then cmd-shift-r to reload the app:

```javascript
localStorage.compass_guide_cues = "[]"
```

Screenshot 2025-09-19 at 16 06 43

### CODE REVIEW

File: packages/compass-components/src/components/drawer-portal.tsx [hunk 1]
```diff
@@ -271,21 +273,50 @@ export const DrawerAnchor: React.FunctionComponent = ({ children }) => {
         return orderB < orderA ? 1 : orderB > orderA ? -1 : 0;
       });
   }, [drawerSectionItems]);
+
+  const [assistantNodes, setAssistantNodes] = useState<
+    Record<string, HTMLButtonElement | undefined>
+  >({});
+
+  useEffect(function () {
```
Comment (lerouxb) [inline packages/compass-components/src/components/drawer-portal.tsx]: This obviously loops forever. Trying to find the "right" way to do this. But it works well enough to demonstrate the that the idea will work at least.

File: packages/compass-assistant/src/compass-assistant-drawer.tsx [hunk 2]
```diff
@@ -68,6 +68,8 @@ export const CompassAssistantDrawer: React.FunctionComponent<{
     );
   }
 
+  const appName = 'Compass'; // TODO
```
Comment (lerouxb) [inline packages/compass-assistant/src/compass-assistant-drawer.tsx]: sigh

File: packages/compass-components/src/components/drawer-portal.tsx [hunk 3]
```diff
@@ -271,21 +273,77 @@ export const DrawerAnchor: React.FunctionComponent = ({ children }) => {
...
+
+        const button = document.querySelector<HTMLButtonElement>(
+          `[data-testid="lg-drawer-toolbar-icon_button-${index}"]`
+        );
+        if (button) {
+          nodes[item.id] = button;
+        } else {
+          // we don't re-render enough times for unit tests to pass and this
+          // forces it to keep re-trying until the node is found
+          if (failedLookupCount < 10) {
```
Comment (lerouxb) [inline packages/compass-components/src/components/drawer-portal.tsx]: There must be a better way to do this? It seems to only affect the tests - in Compass proper the node's there immediately.

Comment (lerouxb) [reply packages/compass-components/src/components/drawer-portal.tsx]: This obviously has the extremely edge case bug where if something somehow replaces that dom node it won't pick it up again 😆 I can go further down this hacky rabbit hole to reset the counter again.. (but I'm going to assume toolbar data changes in that case..)

Comment (gribnoysup) [reply packages/compass-components/src/components/drawer-portal.tsx]: There is indeed no way in React to guarantee that when this effect runs the element you are looking for is there. If you want this to handle all these cases consistently this should be using mutation observer as the code below does. Adding a few frames in between effect and this calculation might also help, but it's still a hack. If this is short lived, then it's probably fine as-is (but just barely, this is really not the code we should keep around for long).

Also counter is not being reset, not sure if that's intentional or not, but currently only 10 retries in the whole app lifecycle is allowed here due to that.

Comment (lerouxb) [reply packages/compass-components/src/components/drawer-portal.tsx]: Switched to using MutationObserver now. I feel like it could probably be a bit more efficient?

File: packages/compass-components/src/components/drawer-portal.tsx [hunk 4]
```diff
@@ -271,21 +273,93 @@ export const DrawerAnchor: React.FunctionComponent = ({ children }) => {
...
       });
   }, [drawerSectionItems]);
+
+  const [assistantNodes, setAssistantNodes] = useState<
+    Record<string, HTMLButtonElement | undefined>
+  >({});
+
+  useLayoutEffect(
+    function () {
+      const drawerEl = document.querySelector('.compass-drawer-anchor');
```
Comment (addaleax) [inline packages/compass-components/src/components/drawer-portal.tsx:283]: Should this be specifying a target? e.g. what if we turn the fragment bit on L335 into a `<div id="...">` and then use 
```suggestion
      const drawerEl = document.querySelector('#div-id .compass-drawer-anchor');
```

so that this component doesn't implicitly require that it's a singleton?

File: packages/compass-components/src/components/drawer-portal.tsx [hunk 5]
```diff
@@ -271,21 +273,93 @@ export const DrawerAnchor: React.FunctionComponent = ({ children }) => {
...
+
+      function check() {
+        const nodes: Record<string, HTMLButtonElement | undefined> = {};
+        for (const [index, item] of toolbarData.entries()) {
+          if (!item.guideCue) {
+            continue;
+          }
+
+          const button = document.querySelector<HTMLButtonElement>(
+            `[data-testid="lg-drawer-toolbar-icon_button-${index}"]`
```
Comment (addaleax) [inline packages/compass-components/src/components/drawer-portal.tsx]: Similar question here

File: packages/compass-components/src/components/drawer-portal.tsx [hunk 6]
```diff
@@ -271,21 +273,93 @@ export const DrawerAnchor: React.FunctionComponent = ({ children }) => {
...
+            if (nodes[id] !== oldNodes[id]) {
+              return nodes;
+            }
+          }
+          return oldNodes;
+        });
+      }
+      check();
+
+      const mutationObserver = new MutationObserver(() => {
```
Comment (addaleax) [inline packages/compass-components/src/components/drawer-portal.tsx:320]: TIL 🙏

File: packages/compass-components/src/components/drawer-portal.tsx (hunk 4 above)
Comment (lerouxb) [reply packages/compass-components/src/components/drawer-portal.tsx:283]: I have the same question. Can't see an easy way to do that - looking at the existing layout effect in the same file that also uses a mutation observer it seemed to have run into the same problem and uses a similar selector which I copied here. I think the problem is we can't pass an id for the drawer layout to use and the drawer portal IS the drawer layout.

We're in control of this, though and we only have one drawer portal and one drawer at a time. It would be a pretty profound change to our app's structure.

File: packages/compass-components/src/components/drawer-portal.tsx (hunk 5 above)
Comment (lerouxb) [reply packages/compass-components/src/components/drawer-portal.tsx]: Similar reasoning to the one above but this one is even more temporary until we get leafygreen to add proper support, then we won't have to look up nodes at all. (Updated the description to reflect this plan.)

File: packages/compass-components/src/components/drawer-portal.tsx (hunk 6 above)
Comment (lerouxb) [reply packages/compass-components/src/components/drawer-portal.tsx:320]: Same. I have no idea if I'm using it correctly or ideally - checking with Sergey.

File: packages/compass-components/src/components/drawer-portal.tsx [hunk 7]
```diff
@@ -271,21 +273,93 @@ export const DrawerAnchor: React.FunctionComponent = ({ children }) => {
         return orderB < orderA ? 1 : orderB > orderA ? -1 : 0;
       });
   }, [drawerSectionItems]);
+
+  const [assistantNodes, setAssistantNodes] = useState<
```
Comment (gribnoysup) [inline packages/compass-components/src/components/drawer-portal.tsx]: Nit: this is not necessarily assistant nodes, right? So maybe a naming like this is a bit better:

```suggestion
  const [toolbarIconNodes, setToolbarIconNodes] = useState<
```

File: packages/compass-components/src/components/drawer-portal.tsx (hunk 4 above)
Comment (gribnoysup) [reply packages/compass-components/src/components/drawer-portal.tsx:283]: Hmm, good point! So while we can make this component work not as a singleton, this currently doesn't depend only on anchor container being targeted with a class, you also need to account for this logic in the state provider and practically speaking we do want it to be the only one, so it's not worth the effort making it work universally like that I think.

We indeed don't enforce this singleton behavior right now and we might want to improve this by adding a runtime check that will throw if more than one anchor is found on the page, worth doing a follow up!

File: packages/compass-components/src/components/drawer-portal.tsx [hunk 8]
```diff
@@ -271,21 +273,93 @@ export const DrawerAnchor: React.FunctionComponent = ({ children }) => {
...
+      function check() {
+        const nodes: Record<string, HTMLButtonElement | undefined> = {};
+        for (const [index, item] of toolbarData.entries()) {
+          if (!item.guideCue) {
+            continue;
+          }
+
+          const button = document.querySelector<HTMLButtonElement>(
+            `[data-testid="lg-drawer-toolbar-icon_button-${index}"]`
+          );
```
Comment (gribnoysup) [inline packages/compass-components/src/components/drawer-portal.tsx]: (Maybe) a nit: while this works, it does rely on order of toolbar data being the same as the rendered elements in the toolbar and (what probably bugs me a bit more 😅) on testids. While we don't really have a perfect way to target these elements and this is temporary, here's a way to make this a bit more specific and bind to the actual content of the toolbar data item:

```suggestion
          const button = drawerEl.querySelector(`button[aria-label="${item.label}"]`);
```

Two things here: you only search inside the drawer element you found above, you use actual item data, and you remove the testid from the functional code. Feel free to ignore if this as brittle as the testid approach

File: packages/compass-components/src/components/drawer-portal.tsx (hunk 6 above)
Comment (gribnoysup) [reply packages/compass-components/src/components/drawer-portal.tsx:320]: You can in theory look directly on changed nodes that you get as part of the observed entries, instead of query selecting what you're looking for, but practically this doesn't really matter much here (sometimes you still need to search through the parent if it was added together with the child you expect to see, so still querying involved) and this approach you went with is more straightforward 👍

File: packages/compass-components/src/components/drawer-portal.tsx (hunk 7 above)
Comment (lerouxb) [reply packages/compass-components/src/components/drawer-portal.tsx]: Makes sense. I should have noticed that. Changing now.

File: packages/compass-components/src/components/drawer-portal.tsx (hunk 6 above)
Comment (lerouxb) [reply packages/compass-components/src/components/drawer-portal.tsx:320]: I'm all about the simplicity. And I'm also very aware that this is hopefully a temporary hack. 🤞

File: packages/compass-components/src/components/drawer-portal.tsx (hunk 8 above)
Comment (lerouxb) [reply packages/compass-components/src/components/drawer-portal.tsx]: That's better. I did consider it but was lazy looking into the DOM for something better to use because I felt this is temporary. But it is one line so I might as well and I've made the change.
